File: chess-game/chess-game.v3/web/client/chess-app/src/app/pages/game/play/play/models/pieces/pawn.ts
import { Cell } from "../board/ cell";
import { Side } from "../side";
import { Piece } from "./piece";
import { PieceType } from "./piece_type";

export class Pawn extends Piece {
    
    override get can_move_over(): boolean {
        return false;
    }

    constructor(piece_id: string, side: Side) {
        super(piece_id, side, PieceType.Pawn);
    }

    override validateMove(from: Cell, to: Cell): boolean {
        const [delta_file, delta_rank] = this.calculateMoveDeltas(from, to);

        return delta_file == 0 && (delta_rank == 1 || delta_rank == 2);
    }
}

File: chess-game/chess-game.v3/chessapp/domain/pieces/pawn.py
from chessapp.domain.pieces.piece import Piece
from chessapp.domain.pieces.piece_type import PieceType
from chessapp.domain.rules.movements.pawn_rule import PawnRule
from chessapp.domain.value_objects.piece_id import PieceId
from chessapp.domain.value_objects.side import Side


class Pawn(Piece):

    def get_points(self):
        return 1

    def __init__(self, piece_id: PieceId, side: Side):
        super().__init__(piece_id, side, PieceType.Pawn)
        self._has_moved = False

    def move(self):
        self._has_moved = True

    def has_moved(self):
        return self._has_moved

    def get_rule(self):
        return PawnRule(self._has_moved)

    def promote(self):
        pass